designs/default-zero-value.md

```diff
@@ -0,0 +1,777 @@
+and removed in Smithy 2.0: `PrimitiveBoolean`, `PrimitiveShort`,
+`PrimitiveInteger`, `PrimitiveLong`, `PrimitiveFloat`, and `PrimitiveDouble`.
+Instead of referring to these shapes, refer to their non-primitive shape
+variants (for example, `Integer`, `Boolean`) and apply the `@default` trait to
+structure members. A warning will be emitted when models refer to these shapes,
+which encourages models to update to use shapes that are forward compatible
+with an eventual Smithy 2.0.
+
+Smithy 2.0 IDL support will be added to the 1.x series of Smithy's Java tooling
+so that both 1.0 and 2.0 models can be loaded, simultaneously, and automatically
+converted from 1.0 to 2.0.
+
+
+## New guidance for code generators
+
+Code generated types for structures SHOULD use the `@default` trait and
+`@required` traits to provide member accessors that always return non-null
+values.
+
+- When the `@default` trait is present, the corresponding accessor always
+  returns a non-null value by defaulting missing members with their zero values.
+- When the `@required` trait is present on a member that targets a shape that
+  has a zero value, the corresponding accessor always returns a non-null value.
+- Smithy implementations in languages like TypeScript that do not provide a kind
+  of constructor or builder to create structures may not be able to set default
+  values, precluding them from being able to treat `@required` and `@default`
+  members as non-null.
+
+
+### Serialization
+
+Members marked with the `@default` trait that are set to their default zero
+values SHOULD NOT be serialized.
+
+- Deserializers MUST tolerate receiving a serialized default value. Failing here
+  would leave end-users with no recourse other than to wait for a client or
+  server to update and fix their implementation. This also accounts for older
+  clients that think a structure member is required, but the service has since
+  transitioned the member to use the `@default` trait.
+- Serializing zero values can cause unintended information disclosure like
+  revealing internal-only members (this is the same reason that null values
+  SHOULD NOT be serialized).
+
+
+### Deserialization
+
+Structure members marked as `@required` are not in and of themselves a guarantee
+that they will be serialized. For example, if a service migrates a member from
+`@required` to `@default`, then older clients MUST gracefully handle the zero
+value of the member being omitted on the wire. In this case, rather than
+failing, a client MUST set the member value to its default zero value. Failing
+to deserialize the structure is a bad outcome because what the service
+perceived as a backward compatible change (i.e., removing the `@required`
+trait) could break previously generated clients. The only recourse for the
+client is to tolerate the missing member and fill in the deterministic default
+value.
+
+
+## Alternatives and trade-offs
+
+## A note on added complexity in Smithy 1.x
```

Comment:
Then NullableIndex removes this complexity from the code generators right?

Reply:
It does, but this is more speaking of the complexity in the spec itself, independent of any implementation of the spec.